lib/hashie/extensions/mash/maintain_original_keys.rb

```diff
@@ -0,0 +1,52 @@
+module Hashie
+  module Extensions
+    module Mash
+      # Overrides the indifferent access of a Mash to maintain keys in the
+      # original format given to the Mash.
+      #
+      # @example
+      #   class MaintainingMash < Hashie::Mash
+      #     include Hashie::Extensions::Mash::MaintainOriginalKeys
+      #   end
+      #
+      #   mash = MaintainingMash.new(:symbol_key => :symbol, 'string_key' => 'string')
+      #   mash.to_hash  #=> { :symbol_key => :symbol, 'string_key' => 'string' }
+      #   mash['string_key'] == mash[:string_key]  #=> true
+      #   mash[:symbol_key] == mash['symbol_key']  #=> true
+      module MaintainOriginalKeys
+        private
+
+        def self.included(descendant)
+          unless descendant <= Hashie::Mash
+            fail ArgumentError, "#{descendant} is not a kind of Hashie::Mash"
+          end
+        end
+
+        # Converts symbol/string keys to their alternative formats, but leaves
+        # other keys alone.
+        #
+        # @param [Object, String, Symbol] key the key to convert.
+        # @return [Object, String, Symbol] the converted key.
+        def convert_key(key)
+          case key
+          when Symbol then key.to_s
+          when String then key.to_sym
+          else key
+          end
+        end
+
+        # Indifferently accesses a key in the Mash.
+        #
+        # @param [Object, String, Symbol] key the key to access.
+        # @return [Object] the value assigned to the key.
+        def indifferent_key(key)
```

Comment:
I thought about this overnight and I'm not sure what you mean. The base `#convert_key` just casts everything to a string, so that would not be the same behavior.

I introduced a new method because I feel that it's much more communicative of what's happening. Currently the Mash `#convert_key` method is serving the purpose of `#indifferent_key`, which is "indifferently access this key" vs. "convert this key into a format that can be used indifferently". So I think there's value in the alias in the base Mash class. What this gives us here is an informative method chain that says exactly what it's doing:

1. When indifferently accessing a key, first check to see if the key you gave me exists in the hash. If so, return it.
2. Otherwise, convert the key in whatever method you use for indifferently accessing keys and give me that.

I just realized that there might be an issue with this logic, because the base case should be to return the key that was given, not the converted key. I'm going to play with some tests to double-check, but I'm very interested to see what you mean. It kept my brain spinning last night because the only way I could think of doing what you suggest would make the code less clear in my mind.

Would you mind showing what you mean?